crates/curp/src/server/storage/wal/remover.rs

```diff
@@ -0,0 +1,181 @@
+use std::{
+    io,
+    path::{Path, PathBuf},
+};
+
+use tokio::{
+    fs::File,
+    io::{AsyncReadExt, AsyncWriteExt},
+};
+
+use super::{
+    segment::WALSegment,
+    util::{get_checksum, get_file_paths_with_ext, is_exist, parse_u64, validate_data, LockedFile},
+};
+
+/// The name of the RWAL file
+const REMOVER_WAL_FILE_NAME: &str = "segments.rwal";
+
+/// Atomic remover of segment files
+///
+/// The remover will firstly creates a write ahead log that stores
+/// the removal information, then it will remove it after the segment
+/// removal has completed.
+pub(super) struct SegmentRemover {
+    /// The WAL path for storing the remove information
+    rwal_path: PathBuf,
+}
+
+impl SegmentRemover {
+    #[allow(clippy::doc_markdown)] // False positive for ASCII graph
+    /// Recover from existing RWAL
+    ///
+    /// * RWAL layout
+    ///
+    ///  |----------+----------+-----+----------+-------------------|
+    ///  | record 0 | record 1 | ... | record n | checksum (sha256) |
+    ///  |----------+----------+-----+----------+-------------------|
+    ///
+    /// * The layout of each record
+    ///
+    ///  0      1      2      3      4      5      6      7      8
+    ///  |------+------+------+------+------+------+------+------|
+    ///  | BaseIndex                                             |
+    ///  |------+------+------+------+------+------+------+------|
+    ///  | SegmentID                                             |
+    ///  |------+------+------+------+------+------+------+------|
+    #[allow(clippy::arithmetic_side_effects)] // won't overflow
+    pub(super) async fn recover(dir: impl AsRef<Path>) -> io::Result<()> {
+        let wal_path = Self::rwal_path(&dir);
+        if !is_exist(&wal_path) {
+            return Ok(());
+        }
+
+        let mut wal = LockedFile::open_rw(wal_path.clone())?.into_async();
+        let mut buf = vec![];
```

Comment:
Nit: is it better to use `bytes::Bytes` to instead of `std::Vec`?

Reply:
`bytes::Bytes` is incompetible with rust std. The following `read_to_end` only support read to a Vec.